src/Symfony/Component/Serializer/Instantiator/InstantiatorInterface.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Instantiator;
+
+use Symfony\Component\Serializer\Exception\MissingConstructorArgumentsException;
+
+/**
+ * @author Jérôme Desjardins <jewome62@gmail.com>
+ */
+interface InstantiatorInterface
+{
+    /**
+     * Instantiates a new object.
+     *
+     * @throws MissingConstructorArgumentsException When some arguments are missing to use the constructor
+     *
+     * @return mixed
+     */
+    public function instantiate(string $class, array &$data, array &$context, string $format = null): object;
```

Comment:
This is actually how it's working today https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php#L329

I didn't found a way to do it better sadly

Reply:
You can create a copy of the array on top of the method, and use the mutable copy inside: this way you ensure you will not modify the incoming array for the rest of world, but still keep the mutable working the same way it does work today inside. You don't have to make it a reference for the rest of the world, else it seems like a very highly magic and undocumented behaviour.

Reply:
We do want to modify it for the "rest of the world", that's the intended behavior.

For example when we have 5 datas in the Normalizer but 2 are set through Instantiator, we remove them from the `$data` array. Like that the ObjectNormalizer won't try to set them again.

Reply:
Thanks for explaining.

Reply:
Also think, it would be better to not share this logic, WDYT of using an intermediate object returned by the interface, something like:

```php

public function instantiate(string $class, array &$data, array &$context, string $format = null): InstantiateResultInterface;

...
interface InstantiateResultInterface {
    // Instantiated object
    public function getObject();
    // Data without used parameters
    public function getUnusedData(): array;
    // New context without object used in it
    public function getUnusedContext(): array;
}
```

Effect of this, is that it will be the normalizer responsability to correctly change its data and context according to the result of the instantiate method (and no more the implementation).

Not sure if its better in terms of DX though...

Also maybe it's not the best thing to make it an Interface, (but having it this way would also allow some implementation to add debugging information to this object which can be used in the profiler)

Reply:
@joelwurtz indeed, I don't see another way to do this this either. And I agree, I'm not sure this would be the best for DX as well.

I don't know why the context itself is a reference as well ? Does the instanciate() method is supposed to modify it and add data as well within ?

Reply:
`$context` is a reference because we remove the `object_to_populate` if we use it, to avoid some other Normalizer to use it.

I will add an object to pass data, context & errors :+1:

Reply:
Ouch, there's much internal bahviours inherited from own normalizer's needs. Yes, returning and object is probably more comprehensible for users, it's auto-documenting at the very least.

Reply:
I just added a `InstantiatorResult` object to store instantiate object, data, context and errors.

Errors was introduce to avoid having blocking exceptions and allow to return a `null` object in `Instantiator`, that way we could have (in future) a `ChainInstantiator` that will manage multiple implementations.

Reply:
It's nice, at least in my opinion it makes the interface more comprehensible, but that's a subjective point of view. I did a quick review it's mostly nitpicks and a few questions.